src/scanner.c

```diff
@@ -0,0 +1,448 @@
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <tree_sitter/parser.h>
+#include <wctype.h>
+
+/**
+ * Debugging helper macros. Example output:
+ *
+ * > HEREDOC_START
+ * scan_start() <-
+ * next E
+ * next O
+ * next F
+ * del EOF
+ * set HEREDOC_START
+ * stop \n
+ * next \n
+ * scan_delimiter() <-
+ * scan_delimiter() -> false
+ * scan_start() -> true
+ *
+ * > HEREDOC_START_NEWLINE HEREDOC_BODY HEREDOC_END_NEWLINE HEREDOC_END
+ * scan_body() <-
+ * next \n
+ * stop $
+ * scan_delimiter() <-
+ * scan_delimiter() -> false
+ * set HEREDOC_START_NEWLINE
+ * scan_body() -> true
+ */
+
+#define debug 1
+
+#define print(...) \
+  if (debug) printf(__VA_ARGS__)
+
+#define peek() lexer->lookahead
+
+#define next()                    \
+  {                               \
+    char *chr = str(peek());      \
+    print("next %s\n", chr);      \
+    lexer->advance(lexer, false); \
+  }
+
+#define skip()                   \
+  {                              \
+    char *chr = str(peek());     \
+    print("skip %s\n", chr);     \
+    lexer->advance(lexer, true); \
+  }
+
+#define stop()               \
+  {                          \
+    char *chr = str(peek()); \
+    print("stop %s\n", chr); \
+    lexer->mark_end(lexer);  \
+  }
+
+#define set(symbol)                        \
+  {                                        \
+    print("set %s\n", TokenTypes[symbol]); \
+    lexer->result_symbol = symbol;         \
+  }
+
+#define ret(function, result)                                   \
+  print("%s() -> %s\n", function, (result) ? "true" : "false"); \
+  return result;
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+enum TokenType {
+  HEREDOC_START,
+  HEREDOC_START_NEWLINE,
+  HEREDOC_BODY,
+  HEREDOC_END_NEWLINE,
+  HEREDOC_END,
+  EMBEDDED_OPENING_BRACE,
+};
+
+const char *const TokenTypes[] = {
+    "HEREDOC_START",           //
+    "HEREDOC_START_NEWLINE",   //
+    "HEREDOC_BODY",            //
+    "HEREDOC_END_NEWLINE",     //
+    "HEREDOC_END",             //
+    "EMBEDDED_OPENING_BRACE",  //
+};
+
+typedef struct {
+  uint32_t cap;
+  uint32_t len;
+  char *data;
+} String;
+
+String string_new() { return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)}; }
```

Comment:
another `sizeof(char)`
also I would make put this on multiple lines

Reply:
i think it's pretty confusing that `cap` is not the actual capacity of the array (it's one less).
But this invariant seems to be held everywhere.

Reply:
it's for the null terminator

Reply:
indeed. as i said, it's a bit confusing but seems to be done well here.
my only actionable comment is the first one.